src/engine/GroupBy.h

```diff
@@ -190,11 +192,97 @@ class GroupBy : public Operation {
                    const sparqlExpression::EvaluationContext* ctx) {
       if (ValueGetter{}(AD_FWD(value), ctx)) count_++;
     }
-    [[nodiscard]] ValueId calculateResult() const {
+    [[nodiscard]] ValueId calculateResult(const LocalVocab* localVocab) const {
+      (void)localVocab;
       return ValueId::makeFromInt(count_);
     }
   };
 
+  // Data to perform MIN/MAX aggregation using the HashMap optimization.
+  template <valueIdComparators::Comparison Comp>
+  struct ExtremumAggregationData {
+    sparqlExpression::IdOrString currentValue_ = ValueId::makeUndefined();
+    void increment(const sparqlExpression::IdOrString& value,
+                   const sparqlExpression::EvaluationContext* ctx) {
+      if (auto val = std::get_if<ValueId>(&currentValue_);
+          val != nullptr && val->isUndefined()) {
+        currentValue_ = value;
+        return;
+      }
+      auto impl = [&ctx]<typename X, typename Y>(const X& x, const Y& y) {
+        return toBoolNotUndef(
+            sparqlExpression::compareIdsOrStrings<
+                Comp, valueIdComparators::ComparisonForIncompatibleTypes::
+                          CompareByType>(x, y, ctx));
+      };
+      if (std::visit(impl, value, currentValue_)) currentValue_ = value;
+    }
+    [[nodiscard]] ValueId calculateResult(LocalVocab* localVocab) const {
+      if (auto val = std::get_if<ValueId>(&currentValue_))
+        return *val;
+      else if (auto str = std::get_if<std::string>(&currentValue_)) {
+        auto localVocabIndex = localVocab->getIndexAndAddIfNotContained(*str);
+        return ValueId::makeFromLocalVocabIndex(localVocabIndex);
+      }
+      return ValueId::makeUndefined();
+    }
+  };
+
+  using MinAggregationData =
+      ExtremumAggregationData<valueIdComparators::Comparison::LT>;
+  using MaxAggregationData =
+      ExtremumAggregationData<valueIdComparators::Comparison::GT>;
+
+  // Data to perform the SUM aggregation using the HashMap optimization.
+  struct SumAggregationData {
+    using ValueGetter = sparqlExpression::detail::NumericValueGetter;
+    bool error_ = false;
+    double sum_ = 0;
+    void increment(auto&& value,
+                   const sparqlExpression::EvaluationContext* ctx) {
+      auto val = ValueGetter{}(AD_FWD(value), ctx);
+
+      if (const int64_t* intval = std::get_if<int64_t>(&val))
+        sum_ += static_cast<double>(*intval);
+      else if (const double* dval = std::get_if<double>(&val))
+        sum_ += *dval;
+      else
+        error_ = true;
+    };
```

Comment:
Maybe also use a safe approach using `std::visit` here in general.
(either using the `OverloadCallOperator` or using `if constexpr` inside the lambda.
You can find examples on cppreference for std::visit or accross the codebase, alternatively contact me if you have trouble.